ConvertToGematriaNumericString with separator is wrong for single chars (10,20,30)

itsho opened the pull request:
currently, ConvertToGematriaNumericString(10) is returning 

> "י

So, in these specific cases (single char when the separator is required), I've added the thousands separator - AFTER the character.
also, I've added the relevant unit-tests.

thanks.

itsho committed:
for specific cases (single char when separator is required), we use the thousands separator - AFTER the character...

itsho committed:
ConvertToGematriaNumericString SingleDigit WithSeparators ReturnsCorrectString

itsho commented:
my second commit takes care of single character conversion (**1 to א'**) when the separator is required.

yellis commented on .gitignore:
```diff
@@ -4,4 +4,5 @@ obj
-*.nupkg
\ No newline at end of file
+*.nupkg
+/.vs/Gematria/v16
```
I dont think that this line is necessary
```suggestion
```

itsho commented:
You are right.
Do you want me to remove it?

________________________________
From: Yaakov Ellis <notifications@github.com>
Sent: Tuesday, October 22, 2019 12:03:13 PM
To: yellis/Gematria <Gematria@noreply.github.com>
Cc: Itamar Shoham <itsho.itsho@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [yellis/Gematria] ConvertToGematriaNumericString with separator is wrong for single chars (10,20,30) (#2)


@yellis commented on this pull request.

________________________________

In .gitignore<https://github.com/yellis/Gematria/pull/2#discussion_r337401428>:

> \ No newline at end of file

+*.nupkg

+/.vs/Gematria/v16



I dont think that this line is necessary

⬇️ Suggested change

-/.vs/Gematria/v16



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/yellis/Gematria/pull/2?email_source=notifications&email_token=AANRJTLRZKO6DZPN777CAX3QP26VDA5CNFSM4JDFOWL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIXMFKI#pullrequestreview-305054377>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AANRJTOGZJRQ7XBALLNB2GTQP26VDANCNFSM4JDFOWLQ>.

yellis commented:
yeah, you should be able to just accept the suggestion in the GH UI

yellis commented:
I am not sure about this. Seems to me that when the number results in one hebrew letter, then there should be no apostrophe or quote appearing. That this separator character is only there to separate multiple gematria chars, to make it clear that they are not a word.

You have seen it done this way elsewhere?

Would be more comfortable with adding in an options flag enum, one of which can be "AddQuoteAfterSingleCharGematria". (Flag enum so that this can be used for other options in the future as well).

itsho committed:
Update .gitignore

Co-Authored-By: Yaakov Ellis <yellis@gmail.com>

itsho committed:
Update .gitignore

Co-Authored-By: Yaakov Ellis <yellis@gmail.com>

itsho commented:
yes. if you'll use the following code for dates of single char (א-י,כ,ל), you'll see that you get the apostrophe at the end.

```
public string GetHebrewDateString(DateTime value)
{
     var ci = CultureInfo.CreateSpecificCulture("he-IL");
     ci.DateTimeFormat.Calendar = _hebCal;
     return value.ToString("d", ci);
}
```

yellis commented:
I hear. Not to sure that MS is the authority on this stuff though...

What do you think about adding a flags enum to control whether or not to do this?

itsho commented:
I have no problem adding the enum. 
your call.

itsho committed:
* moved some the options into a new class - including 'AddQuoteAfterSingleChar'

* added unit-test related to the case:
ConvertToGematriaNumericString_AnyDigitAmount_WithSeparators_WithDoubleQoute_ReturnsCorrectString

itsho committed:
* moved some the options into a new class - including 'AddQuoteAfterSingleChar'

* added unit-test related to the case:
ConvertToGematriaNumericString_AnyDigitAmount_WithSeparators_WithDoubleQoute_ReturnsCorrectString

itsho commented:
I thought about it, and eventually, I think that when using a flag it would force us to add some complex logic to avoid invalid cases (AddQuoteAfterSingleChar | AddDoubleQuoteAfterSingleChar).

So, as you can see in my last commit -  I've used a separate class (**GematriaOptions**) for options, and I've added: 

    public bool AddQuoteAfterSingleChar { get; }

itsho commented:
Hi. are you OK with the changes I did?

thanks,

yellis commented on EllisWeb.Gematria/GematriaOptions.cs:
```diff
@@ -0,0 +1,67 @@
+        /// <summary>
+        /// default value for <see cref="IncludeSeparators"/>
+        /// </summary>
+        public const bool DEFAULT_SHOULD_INCLUDE_SEPARATORS = true;
```
Just set the default for this on the relevant line

```
public bool IncludeSeparators { get; } = true;
```

itsho replied:
and remove the const?
in that case, we are forcing the user to always send all parameters to `GematriaOptions` ctor.

yellis commented on EllisWeb.Gematria/GematriaOptions.cs:
```diff
@@ -0,0 +1,67 @@
+        /// <summary>
+        /// default value for <see cref="ThousandsSeparator"/>
+        /// </summary>
+        public const char DEFAULT_THOUSANDS_SEPARATOR = '\'';
```
default should be set on relevant line

itsho replied:
and remove the const?
in that case, we are forcing the user to always send all parameters to `GematriaOptions` ctor.

yellis commented on EllisWeb.Gematria/GematriaOptions.cs:
```diff
@@ -0,0 +1,67 @@
+        /// <summary>
+        /// default value for <see cref="TensSeparator"/>
+        /// </summary>
+        public const char DEFAULT_TENS_SEPARATOR = '"';
```
default should be set on relevant line

itsho replied:
and remove the const?
in that case, we are forcing the user to always send all parameters to `GematriaOptions` ctor.

yellis commented on EllisWeb.Gematria/GematriaOptions.cs:
```diff
@@ -0,0 +1,67 @@
+        /// <summary>
+        /// default value for <see cref="AddQuoteAfterSingleChar"/>
+        /// </summary>
+        public const bool DEFAULT_ADD_QUOTE_AFTER_SINGLE_CHAR = true;
```
default should be set on relevant line

itsho replied:
and remove the const?
in that case, we are forcing the user to always send all parameters to `GematriaOptions` ctor.

yellis commented on EllisWeb.Gematria/GematriaOptions.cs:
```diff
@@ -0,0 +1,67 @@
+        /// <param name="thousandsSeparator">Value to use separating between thousands-groupings. Defaults to a single quote (')</param>
+        /// <param name="tensSeparator">Value to use separating between the tens and single digit letters. Defaults to a double quote (")</param>
+        /// <param name="addQuoteAfterSingleChar">When the result is a single char, Should we place a single quote (true) or a double-quote (false)</param>
+        public GematriaOptions(bool includeSeparators = DEFAULT_SHOULD_INCLUDE_SEPARATORS,
```
once defaults are set on the relevant property lines, no need to have all of this logic in the constructor

itsho replied:
I think that this is a good thing to have the default parameters.
why force the users to send all parameters every time?

yellis replied:
no parameters on the constructor at all. People can set whatever they want in the class.

    new GematriaOptions { includeSeparators = true };

or

    new GematriaOptions(); // uses defaults for all properties

are both valid. Defaults for properties are used.

yellis commented on EllisWeb.Gematria/Calculator.cs:
```diff
@@ -167,50 +165,89 @@ public static long GetNumericGematriaValue(string sourceString, GematriaType gem
         /// A value of 15 will always be represented as ט"ו and 16 will be represented as ט"ז, following Jewish custom. 
         /// </remarks>
         /// <returns>Gemtria Numeric representation of given number</returns>
-        public static string ConvertToGematriaNumericString(long number, bool includeSeparators = true, char thousandsSeparator = '\'', char tensSeparator = '"')
```
please keep the original signature as an overload, and just generate a `GematriaOptions` object there using the values provided. No need to break this for people unnecessarily.

itsho replied:
ok.

itsho committed:
bring back ConvertToGematriaNumericString original method as overload

itsho committed:
bring back ConvertToGematriaNumericString original method as overload

itsho committed:
GematriaOptions - set default values for the properties

itsho committed:
GematriaOptions - set default values for the properties

yellis reviewed (changes requested):
no need for any parameters in the contructor. And no need for separate `DEFAULT_…` `const` variables. Just set the defaults for each properties

itsho commented:
I'm not following. How would one use different values than the default without having the ctor parameters? All of the properties are public getters with private setters...

yellis commented on EllisWeb.Gematria/Calculator.cs:
```diff
@@ -168,49 +168,110 @@ public static long GetNumericGematriaValue(string sourceString, GematriaType gem
         /// </remarks>
         /// <returns>Gemtria Numeric representation of given number</returns>
         public static string ConvertToGematriaNumericString(long number, bool includeSeparators = true, char thousandsSeparator = '\'', char tensSeparator = '"')
```
please decorate with an `[obsolete]` attribute, and give explanation for this in `remarks`

yellis reviewed (changes requested):
It is a matter of style preference. No right answer. My preference in this case is to:

- public properties
- parameterless contructor
- no `const` defaults
- set defaults on all of the properties

itsho committed:
Did the changes requested by yellis in pull request

* Removed defaults from GematriaOptions
* Marked previous ConvertToGematriaNumericString with [Obsolete]
* Fixed the unit-tests to work correctly with GematriaOptions

yellis commented on EllisWeb.Gematria/Calculator.cs:
```diff
@@ -167,50 +168,116 @@ public static long GetNumericGematriaValue(string sourceString, GematriaType gem
+        /// A value of 15 will always be represented as ט"ו and 16 will be represented as ט"ז, following Jewish custom. 
+        /// </remarks>
+        /// <returns>Gemtria Numeric representation of given number</returns>
+        public static string ConvertToGematriaNumericString(long number, GematriaOptions options)
```
Should change this to 

`public static string ConvertToGematriaNumericString(long number, GematriaOptions options = null)`

And then the first line should be `options = options ?? new GematriaOptions();`

That way it is not required to give options, and if you dont, it will use the defaults.

itsho committed:
ConvertToGematriaNumericString - options should be null by default.

itsho commented:
Thanks!
may I ask if you planning to release a new NuGet package sometime soon?

Thanks again.

yellis commented:
@itsho [done](https://www.nuget.org/packages/Gematria/1.1.0)

itsho commented:
Thank you
